src/streaming/FetchLoader.js

```diff
@@ -0,0 +1,181 @@
+ *  contributors may be used to endorse or promote products derived from this software
+ *  without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS AS IS AND ANY
+ *  EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ *  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ *  IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ *  INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ *  NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ *  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ *  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import FactoryMaker from '../core/FactoryMaker';
+import ISOBoxer from 'codem-isoboxer';
+
+/**
+* @module FetchLoader
+* @description Manages download of resources via HTTP using fetch.
+* @param {Object} cfg - dependancies from parent
+*/
+function FetchLoader(cfg) {
+
+    cfg = cfg || {};
+
+    let instance;
+
+    function send(httpRequest) {
+
+        // Variables will be used in the callback functions
+        let firstProgress = true; /*jshint ignore:line*/
+        let needFailureReport = true; /*jshint ignore:line*/
+        let requestStartTime = new Date();
+        let lastTraceTime = requestStartTime; /*jshint ignore:line*/
+        let lastTraceReceivedCount = 0; /*jshint ignore:line*/
+
+        let request = httpRequest.request;
+
+        const headers = new Headers(); /*jshint ignore:line*/
+        if (request.range) {
+            headers.append('Range', 'bytes=' + request.range);
+        }
+
+        if (!request.requestStartDate) {
+            request.requestStartDate = requestStartTime;
+        }
+
+        const reqOptions = {
+            method: httpRequest.method,
+            headers: headers,
+            credentials: httpRequest.withCredentials ? 'include' : undefined
+        };
+
+        fetch(httpRequest.url, reqOptions).then(function (response) {
+            if (!response.ok) {
+                httpRequest.onend();
+            }
+
+            if (!httpRequest.response) {
+                httpRequest.response = {};
+            }
+            httpRequest.response.status = response.status;
+            httpRequest.response.statusText = response.statusText;
+            httpRequest.response.responseURL = response.url;
+            let responseHeaders = '';
+            for (const key of response.headers.keys()) {
+                responseHeaders += key + ': ' + response.headers.get(key) + '\n';
+            }
+            httpRequest.response.responseHeaders = responseHeaders;
+
+            const totalBytes = parseInt(response.headers.get('Content-Length'), 10);
```

Comment:
This doesn't really make sense. When using chunked transfer encoding, the content length is omitted as it is unknown, so `totalBytes` will always be `NaN`.

Reply:
I will take a look at this part. There is just one case in which totalBytes could be higher than 0, that is when a user  enables Low Latency Mode but the HTTP server doesn't support chunked transfer encoding. In those cases we have two options:

- Detect that situation and disable low latency mode after the first request. More complex to deal with (what we do in situations in which segments could be served from more than one server/cdn? one could support chunked transfer encoding while the other no)
- Continue using fetch and mimic the behavior of xhr loader.

Reply:
I don't think you should disable low latency mode if the application has set it. That leads to awkward loss of state. If the server doesn't support chunked transfer of the chunked-encoded content, then there is no harm in continuing to use the exact same fetch based routine to retrieve the data. The difference will be that instead of the data streaming in at a relatively constant rate, there will be nothing and then a sudden dump of data at the completion of the segment (analogous to what we get with XHR today).

Reply:
> analogous to what we get with XHR today

Agreed, and in which case, why bother switching out XHRLoader and FetchLoader at all? On supported platforms, it seems we should just use Fetch all the time (assuming the Loaders have feature parity)?

Reply:
I see this as a temporary solution. Low Latency support (fetch included) is going to be released in experimental mode (need more tests, feedback from community, etc).  It is a change that affects other parts of dash.js (for example, required us to modify the way we interact with SourceBuffers, it will affect the ABR algorithm) and don't want to apply it without being sure it is not breaking anything and it is not impacting support for any browser/device. 

Once we feel confortable, I agree with you. We could use Fetch as the primary way for doing requests and use XHR just as a fallback mechanism in case the device/browser doesn't support fetch.